Kashoo/synctos
# Issue 282: Allow midnight to be represented as hour 24

# Description

Provides parity with the ECMAScript specification for date and time strings by allowing a time of "24:00:00.000"/"24:00:00"/"24:00" to represent midnight in the `datetime` and `time` validation types.

# Testing

For the `datetime` validation type:

1. Generate a sync function from `datetime-doc-definitions.js`: `./make-sync-function -j test/resources/datetime-doc-definitions.js build/output.txt`
2. Insert the generated sync function into a Sync Gateway configuration file
3. Add/replace a document with an ID of "datetimeDoc" and `formatValidationProp` property values of "2018-04-06T24:00:00.000-07:00", "2018-04-06T24:00:00-07:00" and "2018-04-06T24:00-07:00". They should all be accepted.
4. Attempt to add/replace a document with an ID of "datetimeDoc" and `formatValidationProp` property values of "2018-04-06T24:00:00.001-07:00", "2018-04-06T24:00:01-07:00" and "2018-04-06T24:01-07:00". They should all be rejected as invalid date-times.

For the `time` validation type:

1. Generate a sync function from `time-doc-definitions.js`: `./make-sync-function -j test/resources/time-doc-definitions.js build/output.txt`
2. Insert the generated sync function into a Sync Gateway configuration file
3. Add/replace a document with an ID of "timeDoc" and `formatValidationProp` property values of "24:00:00.0", "24:00:00" and "24:00". They should all be accepted.
4. Attempt to add/replace a document with an ID of "timeDoc" and `formatValidationProp` property values of "24:00:00.1", "24:00:01" and "24:01". They should all be rejected as invalid times.

# Related Issue

* GitHub issue link: #282
